changefeedccl: add backfill progress metric

samiskin opened the pull request:
Previously there was no way of knowing how far along a backfill was in
its task.  This made it hard to debug issues.

This patch adds a "changefeed.backfill_pending_ranges" SLI metric to allow
approximating backfill progress using the number of ranges left to
export.  This value is cleared if the changefeed is cancelled.

Release note (ops change): The "changefeed.backfill_pending_ranges"
prometheus metric was added to track ongoing backfill progress of a
changefeed.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/76995)
<!-- Reviewable:end -->

samiskin committed:
changefeedccl: add backfill progress metric

Previously there was no way of knowing how far along a backfill was in
its task.  This made it hard to debug issues.

This patch adds a "changefeed.backfill_pending_ranges" SLI metric to allow
approximating backfill progress using the number of ranges left to
export.  This value is cleared if the changefeed is cancelled.

Release note (enterprise change): The
"changefeed.backfill_pending_ranges" prometheus metric was added to
track ongoing backfill progress of a changefeed.

Release justification: Low risk metrics changes

miretskiy reviewed (changes requested):
> Reviewed 1 of 4 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76995)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @samiskin and @stevendanna)

___
*[-- commits, line 11 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76995#-Mx0k7xjDemVrxUyp9Gx):*
should the category be `enterprise change`?
___
*[pkg/ccl/changefeedccl/metrics.go, line 162 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76995#-Mx0nlMw3fJ3xHlbuhHM:-Mx0nlMw3fJ3xHlbuhHN:b4csgl2) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b02ede529dbb2455185b46ef9bf6bfa5930b1df/pkg/ccl/changefeedccl/metrics.go#L162)):*
> ```Go
> 		var mu syncutil.Mutex
> 		m.BackfillPendingRanges.Inc(initial)
> 		remaining := initial
> ```

I think the use of atomics is justified here...
___
*[pkg/ccl/changefeedccl/metrics.go, line 167 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76995#-Mx0nqsu7TVjr2GAaGLE:-Mx0nqsu7TVjr2GAaGLF:b-e8uhm5) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b02ede529dbb2455185b46ef9bf6bfa5930b1df/pkg/ccl/changefeedccl/metrics.go#L167)):*
> ```Go
> 			mu.Lock()
> 			defer mu.Unlock()
> 			if remaining > 0 {
> ```

wouldn't this be a bug?
Anyways, I think you should just decrement BackfillPendingRanges, and also atomic decrement remaining.
___
*[pkg/ccl/changefeedccl/metrics.go, line 177 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76995#-Mx0mkvw3koe-OoyoDNR:-Mx0mkvw3koe-OoyoDNS:bbz7gvs) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b02ede529dbb2455185b46ef9bf6bfa5930b1df/pkg/ccl/changefeedccl/metrics.go#L177)):*
> ```Go
> 			defer mu.Unlock()
> 			m.BackfillPendingRanges.Inc(-remaining)
> 			remaining = 0
> ```

drop remaining = 0? 
___
*[pkg/ccl/changefeedccl/kvfeed/scanner.go, line 104 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76995#-Mx0kpev6q2gFhc5RL8X:-Mx0kpev6q2gFhc5RL8Y:bvxxamm) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b02ede529dbb2455185b46ef9bf6bfa5930b1df/pkg/ccl/changefeedccl/kvfeed/scanner.go#L104)):*
> ```Go
> 			if err == nil {
> 				backfillDec()
> 			}
> ```

I think you should just `defer backfillDec()` above -- this backfill is done, whether it's done with an error or not is not important.



_Code quote:_
```Go
			if err == nil {
				backfillDec()
			}
```
___
*[pkg/ccl/changefeedccl/changefeed_test.go, line 809 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76995#-Mx0oHHqBEIh4BIRwtFK:-Mx0oHHqBEIh4BIRwtFL:bpabv3d) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b02ede529dbb2455185b46ef9bf6bfa5930b1df/pkg/ccl/changefeedccl/changefeed_test.go#L809)):*
> ```Go
> 	t.Run("enterprise", enterpriseTest(testFn, feedTestNoTenants))
> 	t.Run("cloudstorage", cloudStorageTest(testFn, feedTestNoTenants))
> 	t.Run("kafka", kafkaTest(testFn, feedTestNoTenants))
> ```

Probably doesn't need to run against many sinks since this is a metrics specific test;
Also, please make sure this test is not flaky (I don't think it is -- just want to make sure it passes
w/ e.g. `test.count=250`).


<!-- Sent from Reviewable.io -->

samiskin reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76995)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @samiskin, and @stevendanna)

___
*[-- commits, line 11 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76995#-Mx0k7xjDemVrxUyp9Gx):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

should the category be `enterprise change`?
</blockquote></details>

👍 I wasn't sure if metrics-only counted as operations.  I guess anything changefeed-related is an enterprise change?
___
*[pkg/ccl/changefeedccl/metrics.go, line 162 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76995#-Mx0nlMw3fJ3xHlbuhHM:-Mx0oyV8-u5QnDlfNKLP:bjho2s5) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b02ede529dbb2455185b46ef9bf6bfa5930b1df/pkg/ccl/changefeedccl/metrics.go#L162)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I think the use of atomics is justified here...
</blockquote></details>

The reason I just used a mutex was that I ended up having compare and swap loops:
```go
dec = func() {
	var oldRem int64
	for { // Consume one of the remaining units
		oldRem = atomic.LoadInt64(&remaining)
		if oldRem <= 0 || atomic.CompareAndSwapInt64(&remaining, oldRem, oldRem-1) {
			break
		}
	}
	if oldRem > 0 {
		m.BackfillPendingRanges.Dec(1)
	}
}

clear = func() {
	for { // Consume all of the remaining units
		oldRem := atomic.LoadInt64(&remaining)
		if atomic.CompareAndSwapInt64(&remaining, oldRem, 0) {
			m.BackfillPendingRanges.Dec(remaining)
			break
		}
	}
}

```
Since I was worried that you'd have one thread loading in a non-0 value, the next clears the metric to 0, then the first thread decreases the metric now to -1.

If its still better to have the loops then I can swap it back.

Technically we should never have a clear() during an ongoing dec() since I only call clear() once the backfill completes, but I liked the idea of covering that case regardless to not implicitly couple this metrics code with how it happens to be called right now.

I could also eliminate this but add a comment of "Ensure that clear() is not called while dec() is still able to be called" so that the dependency is at least explicit.
___
*[pkg/ccl/changefeedccl/metrics.go, line 167 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76995#-Mx0nqsu7TVjr2GAaGLE:-Mx0qXCUBjDt2Iv4aRIo:b-mn8sxq) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b02ede529dbb2455185b46ef9bf6bfa5930b1df/pkg/ccl/changefeedccl/metrics.go#L167)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

wouldn't this be a bug?
Anyways, I think you should just decrement BackfillPendingRanges, and also atomic decrement remaining.
</blockquote></details>

With the current code that would be a bug, but similar to my earlier point, this is covering the "dec() after clear()" case just to avoid .
___
*[pkg/ccl/changefeedccl/metrics.go, line 177 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76995#-Mx0mkvw3koe-OoyoDNR:-Mx0sairBOrxX7DIFdee:b-6alh8t) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b02ede529dbb2455185b46ef9bf6bfa5930b1df/pkg/ccl/changefeedccl/metrics.go#L177)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

drop remaining = 0? 
</blockquote></details>

This is also covering the "dec() after clear()"
___
*[pkg/ccl/changefeedccl/kvfeed/scanner.go, line 104 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76995#-Mx0kpev6q2gFhc5RL8X:-Mx0szEF5gFQUGzCrK1x:b-mt5jdh) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b02ede529dbb2455185b46ef9bf6bfa5930b1df/pkg/ccl/changefeedccl/kvfeed/scanner.go#L104)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I think you should just `defer backfillDec()` above -- this backfill is done, whether it's done with an error or not is not important.
</blockquote></details>

I initially added this check because it made the unit test easier, but I also liked the idea that if there were errors on certain ranges during the backfill you'd be able to get an idea of how many failed based on how sharply the count dropped at the end.


<!-- Sent from Reviewable.io -->

samiskin reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76995)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @samiskin, and @stevendanna)

___
*[pkg/ccl/changefeedccl/metrics.go, line 162 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76995#-Mx0nlMw3fJ3xHlbuhHM:-Mx0v-po3OPmi-uFTqHu:b4knl7m) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b02ede529dbb2455185b46ef9bf6bfa5930b1df/pkg/ccl/changefeedccl/metrics.go#L162)):*
<details><summary><i>Previously, samiskin (Shiranka Miskin) wrote…</i></summary><blockquote>

The reason I just used a mutex was that I ended up having compare and swap loops:
```go
dec = func() {
	var oldRem int64
	for { // Consume one of the remaining units
		oldRem = atomic.LoadInt64(&remaining)
		if oldRem <= 0 || atomic.CompareAndSwapInt64(&remaining, oldRem, oldRem-1) {
			break
		}
	}
	if oldRem > 0 {
		m.BackfillPendingRanges.Dec(1)
	}
}

clear = func() {
	for { // Consume all of the remaining units
		oldRem := atomic.LoadInt64(&remaining)
		if atomic.CompareAndSwapInt64(&remaining, oldRem, 0) {
			m.BackfillPendingRanges.Dec(remaining)
			break
		}
	}
}

```
Since I was worried that you'd have one thread loading in a non-0 value, the next clears the metric to 0, then the first thread decreases the metric now to -1.

If its still better to have the loops then I can swap it back.

Technically we should never have a clear() during an ongoing dec() since I only call clear() once the backfill completes, but I liked the idea of covering that case regardless to not implicitly couple this metrics code with how it happens to be called right now.

I could also eliminate this but add a comment of "Ensure that clear() is not called while dec() is still able to be called" so that the dependency is at least explicit.
</blockquote></details>

This could also happen in the concurrent `dec()` case but it would be due to a bug that resulted in more calls to `dec()` than the initial amount.  It could be nice to have this even in that case since then at least the bug wouldn't permanently skew the metric.


<!-- Sent from Reviewable.io -->

miretskiy commented:
> Since I was worried that you'd have one thread loading in a non-0 value, the next clears the metric to 0, then the first thread decreases the metric now to -1.

I think where you're going wrong is that you're trying to Load anything.  All you need to do is to `atomic.AddInt64(&remaining, -1)` in the decrement function, and `m.BackfillPendingRanges.Dec(remaining)` in the clear function.
(You might need to Dec(LoadInt64) instead, but I doubt it because when clear backfill is called, you know that all of the go routines performing backfill have completed; so you only have 1 thread -- and thus, no need to synchronize anything). You can check by running test with `-race`).

miretskiy commented:
> This could also happen in the concurrent `dec()` case but it would be due to a bug that resulted in more calls to `dec()` than the initial amount. It could be nice to have this even in that case since then at least the bug wouldn't permanently skew the metric.

Bugs are, of course, possible.  But in that case we would just produce negative/incorrect values for this gauge; and.. you already have a test; so, maybe let's err on the side of fast and simple code?

miretskiy commented:
> I initially added this check because it made the unit test easier, but I also liked the idea that if there were errors on certain ranges during the backfill you'd be able to get an idea of how many failed based on how sharply the count dropped at the end.

If you are concerned about this and want to track this (though, I doubt we'd need it tbh), a better way would be to have range callback take an error.  And then have 2 metrics -- one for errors.

miretskiy commented on pkg/ccl/changefeedccl/metrics.go:
```diff
@@ -149,6 +152,27 @@ func (m *sliMetrics) getBackfillCallback() func() func() {
+		}
+		clear = func() {
+			atomic.AddInt64(&remaining, -remaining)
+			m.BackfillPendingRanges.Dec(remaining)
```
That doesn't look right.  I don't think you want to update remaining prior to Dec, do you?

miretskiy reviewed (approved):
Approving; but please fix the bug in clear() function.

samiskin commented:
bors r+
